File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/model/List.kt
package com.nouprax.markdown.core

public class List internal constructor(
    public val flavor: ListFlavor,
    public val start: Long?,
    public val variant: OrderedListVariant?,
    public val delimiter: OrderedListDelimiter?,
    public val tight: Boolean,
    public val items: kotlin.collections.List<ListItem>,
    override val scope: Scope,
    override val anchor: String?,
    override val attributes: Attributes,
) : Markup {
    override fun <Result> accept(visitor: Visitor<Result>): Result = visitor.visit(this)
}

public class ListItem internal constructor(
    public val marker: String?,
    public val content: kotlin.collections.List<Markup>,
    override val scope: Scope,
    override val anchor: String?,
    override val attributes: Attributes,
) : Markup {
    public val tasked: Boolean get() = marker != null
    public val completed: Boolean get() = marker != null && marker != " "

    override fun <Result> accept(visitor: Visitor<Result>): Result = visitor.visit(this)
}

File: packages/es-markdown-core/src/model/list.ts
import type { ListFlavor, OrderedListDelimiter, OrderedListVariant } from "../values.js";
import type { MarkupBase } from "./base.js";
import type { Markup } from "./markup.js";

export interface List extends MarkupBase<"list"> {
    readonly flavor: ListFlavor;
    readonly start: number | null;
    readonly variant: OrderedListVariant | null;
    readonly delimiter: OrderedListDelimiter | null;
    readonly tight: boolean;
    readonly items: readonly ListItem[];
}

export interface ListItem extends MarkupBase<"listItem"> {
    readonly marker: string | null;
    readonly exampleLabel: string | null;
    readonly tasked: boolean;
    readonly completed: boolean;
    readonly content: readonly Markup[];
}
